File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.py
def find(start, end, node, left, right):
    global min_value
    global max_value

    if start > right or end < left:
        return

    if start <= left and end >= right:
        min_value = min(min_value, min_tree[node])
        max_value = max(max_value, max_tree[node])
        return

    mid = (left + right) // 2
    find(start, end, node * 2, left, mid)
    find(start, end, node * 2 + 1, mid + 1, right)


from math import log2

MAX = 1000000001

N, M = map(int, input().split())
size = 2 ** int(log2(N - 1) + 1)
min_tree = [MAX] * (size * 2)
max_tree = [0] * (size * 2)

# init
for idx in range(N):
    number = int(input())
    min_tree[size + idx] = number
    max_tree[size + idx] = number

    cur_idx = size + idx

    while cur_idx:
        cur_idx //= 2
        min_tree[cur_idx] = min(min_tree[cur_idx * 2], min_tree[cur_idx * 2 + 1])
        max_tree[cur_idx] = max(max_tree[cur_idx * 2], max_tree[cur_idx * 2 + 1])

# find
for _ in range(M):
    a, b = map(int, input().split())

    min_value = MAX
    max_value = 0
    find(a, b, 1, 1, size)
    print(min_value, max_value)

File: 백준/Gold/2357. 최솟값과 최댓값/최솟값과 최댓값.cc
#include <iostream>
#include <cmath>
using namespace std;

#define min(a, b) (a) > (b) ? (b) : (a)
#define max(a, b) (a) > (b) ? (a) : (b)

int min_tree[1000000];
int max_tree[1000000];

int max_value, min_value;

int find(int start, int end, int node, int left, int right) {
	if ((start > right) || (end < left)) {
		return 0;
	}

	if ((start <= left) && (end >= right)) {
		min_value = min(min_value, min_tree[node]);
		max_value = max(max_value, max_tree[node]);
		return 0;
	}

	int mid = (left + right) / 2;
	find(start, end, node * 2, left, mid);
	find(start, end, node * 2 + 1, mid + 1, right);

	return 0;
}

int power_2(int times) {
	int result = 1;

	while (times > 0) {
		result *= 2;
		times -= 1;
	}

	return result;
}

int main(void) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int N, M, temp, a, b;
	cin >> N >> M;
	
	int log = ceil(log2(N));
	int size = power_2(log);

	for (int i = 0; i < size * 2; i++)
	{
		min_tree[i] = 1000000001;
	}

	for (int i = 0; i < N; i++)
	{
		cin >> temp;
		
		min_tree[i + size] = temp;
		max_tree[i + size] = temp;

		int cur_i = i + size;

		while (cur_i > 0) {
			cur_i /= 2;
			min_tree[cur_i] = min(min_tree[cur_i * 2], min_tree[cur_i * 2 + 1]);
			max_tree[cur_i] = max(max_tree[cur_i * 2], max_tree[cur_i * 2 + 1]);
		}
	}

	for (int i = 0; i < M; i++)
	{
		max_value = 0;
		min_value = 10000000001;

		cin >> a >> b;
		find(a, b, 1, 1, size);

		cout << min_value << " " << max_value << "\n";
	}
}